flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/FullyFinishedOperatorState.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint;
+
+import org.apache.flink.runtime.jobgraph.OperatorID;
+
+/**
+ * A special operator state implementation representing the operators whose instances are all
+ * finished.
+ */
+public class FullyFinishedOperatorState extends OperatorState {
```

Comment:
Should this class also override `equals`/`hashCode`?

And why do we allow to set coordinator state (i.e. don't override like `putState`)?
Can finished vertex have coordinator state?

Reply:
We should indeed need to override equals / hashcode so we could distinguish the finished operator and the empty operator state.  I added these methods.

Regarding coordinator state, currently I also do not have specific examples that keep coordinator state after all the operators are finished. For fully finished operators, we would directly skip their execution after failover, thus all its state should not take effect. Now I also disabled setting the coordinator state. Very thanks for pointing out these issues !